### PR TITLE
Include slashing in block

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

# Description
Add slashings in blocks for proposers

blocked: following the [new design](https://hackmd.prylabs.network/MPXnOpCIQ7GXfATH_zTN1Q) i will first change the rpc structure and then make it ready again

### CODE REVIEW

File: beacon-chain/operations/slashings/service.go [hunk 1]
```diff
@@ -193,13 +193,5 @@ func (p *Pool) validatorSlashingPreconditionCheck(
 	if validator.ExitEpoch() < helpers.CurrentEpoch(state) {
 		return false, nil
 	}
-	// Checking if the validator has already been slashed.
```
Comment (0xKiwi) [inline beacon-chain/operations/slashings/service.go]: Why was this removed?

File: beacon-chain/rpc/service.go [hunk 2]
```diff
@@ -320,7 +341,10 @@ func (s *Service) Stop() error {
 		log.Debug("Initiated graceful stop of gRPC server")
 	}
 	if s.slasherConn != nil {
-		s.slasherConn.Close()
+		if err := s.slasherConn.Close(); err != nil {
+			log.Error(err)
+			return err
```
Comment (0xKiwi) [inline beacon-chain/rpc/service.go]: You should either return the error or log it out, I don't think you should ever be doing both.

File: beacon-chain/rpc/slasher/client.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+func (s *Client) updatePool(ctx context.Context) error {
+	state, err := s.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not get head state: %v", err)
+	}
+
+	if s.SlasherClient != nil {
+		if err := s.updateSlashingPool(ctx, state); err != nil {
+			log.WithError(err)
```
Comment (0xKiwi) [inline beacon-chain/rpc/slasher/client.go]: Refrain from using `log.WithError()`, it doesn't log or print out the error at all, use `log.Error()` instead.
Also, you should only return the error when you should end the current process, and log it out if the process should continue regardless of the error.

File: beacon-chain/rpc/slasher/client.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+			return err
+		}
+	} else {
+		err := status.Error(codes.Internal, "Slasher server has not been started")
+		log.WithError(err)
```
Comment (0xKiwi) [inline beacon-chain/rpc/slasher/client.go]: Same here with the error

File: beacon-chain/rpc/slasher/client.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not retrieve proposer slashings: %v", err)
+	}
+	asr, err := s.SlasherClient.AttesterSlashings(ctx, &slashpb.SlashingStatusRequest{Status: slashpb.SlashingStatusRequest_Active})
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not retrieve attester slashings: %v", err)
+	}
+	for _, ps := range psr.ProposerSlashing {
+		if err := s.SlashingPool.InsertProposerSlashing(state, ps); err != nil {
+			log.WithError(errors.Wrap(err, "could not insert proposer slashing into pool"))
```
Comment (0xKiwi) [inline beacon-chain/rpc/slasher/client.go]:
```suggestion
			log.Error(errors.Wrap(err, "could not insert proposer slashing into pool"))
```

File: beacon-chain/operations/slashings/service.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/operations/slashings/service.go]: slashed vector in state is not a validator ids that were slashed
it is gwei that was slashed at each epoch

File: beacon-chain/rpc/service.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/rpc/service.go]: thanks

**Comment by rauljordan** [comment]
Out of date with the new work being done on slasher in #4836
